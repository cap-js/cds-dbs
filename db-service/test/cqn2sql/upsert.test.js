'use strict'
const cds = require('@sap/cds/lib')
const cqn2sql = require('../../lib/cqn2sql')

beforeAll(async () => {
  cds.model = await cds.load(__dirname + '/testModel').then(cds.linked)
})

describe('upsert', () => {
  test('test with keys only', () => {
    const cqnUpsert = {
      UPSERT: {
        into: 'Foo2',
        columns: ['ID'],
        rows: [[1], [9]],
      },
    }

    const { sql, entries } = cqn2sql(cqnUpsert)
    expect({ sql, entries }).toMatchSnapshot()
  })

  test('test with entries', () => {
    const cqnInsert = {
      UPSERT: {
        into: 'Foo2',
        entries: [
          { ID: 1, name: null, a: 2 },
          { ID: null, name: "'asd'", a: 6 },
        ],
      },
    }

    const { sql, entries } = cqn2sql(cqnInsert)
    expect({ sql, entries }).toMatchSnapshot()
  })
})
