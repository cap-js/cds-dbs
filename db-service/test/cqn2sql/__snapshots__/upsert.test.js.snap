// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upsert test with entries 1`] = `
{
  "entries": [
    [
      "[{"ID":1,"name":null,"a":2},{"ID":null,"name":"'asd'","a":6}]",
    ],
  ],
  "sql": "INSERT INTO Foo2 (ID,name,a) SELECT value->>'$."ID"',value->>'$."name"',value->>'$."a"' FROM json_each(?) WHERE true ON CONFLICT(ID) DO UPDATE SET name = excluded.name,a = excluded.a",
}
`;

exports[`upsert test with keys only 1`] = `
{
  "entries": [
    [
      "[[1],[9]]",
    ],
  ],
  "sql": "INSERT INTO Foo2 (ID) SELECT value->>'$[0]' FROM json_each(?) WHERE true ON CONFLICT(ID) DO NOTHING",
}
`;
